services/language_translation/v2.js

```diff
@@ -21,358 +22,364 @@ module.exports = function (RED) {
+      // Now that the do functions have been defined, can now determine what action this node
+      // is configured for. 
+
       var message = '';
+
       if (!msg.payload) {
         message = 'Missing property: msg.payload';
-        node.error(message, msg)
+        node.error(message, msg);
         return;
       }
 
-      var action = config.action;
-
-      if (!action) {
-        node.warn('Missing action, please select one');
-        node.send(msg);
-        return;
-      }
+      // The dynamic nature of this node has caused problems with the password field. it is 
+      // hidden but not a credential. If it is treated as a credential, it gets lost when there
+      // is a request to refresh the model list. 
+      //
+      // Credentials are needed for each of the modes.
 
-      username = username || this.credentials.username;
-      password = password || this.credentials.password;
+      username = sUsername || this.credentials.username;
+      password = sPassword || this.credentials.password || config.password;
 
       if (!username || !password) {
-        this.status({
-          fill: 'red',
-          shape: 'ring',
-          text: 'missing credentials'
-        });
         message = 'Missing Language Translation service credentials';
-        node.error(message, msg)
+        node.error(message, msg);
         return;
       }
 
-      language_translation = watson.language_translation({
-        username: username,
-        password: password,
-        version: 'v2'
-      });
-
-      switch (action) {
-        case 'translate':
-          this.doTranslate(msg);
-          break;
-        case 'train':
-          this.doTrain(msg);
-          break;
-        case 'getstatus':
-          this.doGetStatus(msg, trainid);
-          break;
-        case 'delete':
-          this.doDelete(msg, trainid);
-          break;
-      }
-    });
-
-    this.doTranslate = function (msg) {
-      var domain = msg.domain || config.domain;
-
-      if (!domain) {
-        node.warn('Missing translation domain, message not translated');
+      var action = msg.action || config.action;
+      if (!action) {
+        node.warn('Missing action, please select one');
         node.send(msg);
         return;
       }
 
-      var srclang  = ''; 
-      var destlang = '';
-
-      switch (domain) {
-        case 'news':
-          srclang = config.srclangnews;
-
-          if (!srclang) {
-            node.warn('Missing source language, message not translated');
-            node.send(msg);
-            return;
-          }
-
-          destlang = config.destlangnews;
-
-          if (!destlang) {
-            node.warn('Missing target language, message not translated');
+      // We have credentials, and know the mode. Further required fields checks
+      // are specific to the requested action.
+      // The required fields are checked, before the relevant function is invoked.
+      switch (action) {
+        case 'translate':
```

Comment:
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [Expected indentation of 8 space characters but found 10.](https://www.codacy.com/app/rezgui-y/node-red-node-watson/file/2613386480/issues/source?bid=3225115&fileBranchId=3311840#l310)